src/main/java/com/microsoft/lst_bench/input/BenchmarkObjectFactory.java

```diff
@@ -90,125 +91,222 @@ private static SparkConnectionManager sparkConnectionManager(
   }
 
   private static PhaseExec createPhaseExec(
       Phase phase,
-      Map<String, TaskTemplate> idToTaskTemplate,
+      InternalLibrary internalLibrary,
       ExperimentConfig experimentConfig,
       Map<String, Integer> taskTemplateIdToPermuteOrderCounter,
       Map<String, Integer> taskTemplateIdToParameterValuesCounter) {
-    List<SessionExec> sessions = new ArrayList<>();
-    for (int i = 0; i < phase.getSessions().size(); i++) {
-      Session session = phase.getSessions().get(i);
+    List<Session> sessions;
+    if (phase.getSessions() != null) {
+      sessions = phase.getSessions();
+    } else if (phase.getTemplateId() != null) {
+      PhaseTemplate phaseTemplate =
+          internalLibrary.getIdToPhaseTemplate().get(phase.getTemplateId());
+      if (phaseTemplate == null) {
+        throw new IllegalArgumentException("Unknown phase template id: " + phase.getTemplateId());
+      }
+      sessions = phaseTemplate.getSessions();
+    } else {
+      throw new IllegalArgumentException("Unknown phase type");
+    }
+    validateSessions(sessions);
+    List<SessionExec> sessionExecList = new ArrayList<>();
+    for (int i = 0; i < sessions.size(); i++) {
+      Session session = sessions.get(i);
       String sessionId = String.valueOf(i);
       SessionExec sessionExec =
           createSessionExec(
               sessionId,
               session,
-              idToTaskTemplate,
+              internalLibrary,
               experimentConfig,
               taskTemplateIdToPermuteOrderCounter,
               taskTemplateIdToParameterValuesCounter);
-      sessions.add(sessionExec);
+      sessionExecList.add(sessionExec);
+    }
+    return ImmutablePhaseExec.of(phase.getId(), sessionExecList);
+  }
+
+  /**
+   * Validates that the sessions have exactly one of template ID, list of tasks, or instance ID
+   * defined.
+   */
+  private static void validateSessions(List<Session> sessions) {
+    for (Session session : sessions) {
+      boolean onlyOneTrue = session.getTemplateId() != null ^ session.getTasks() != null;
+      if (!onlyOneTrue) {
+        throw new IllegalArgumentException(
+            "Must have exactly one of template id or list of tasks defined");
+      }
     }
-    return ImmutablePhaseExec.of(phase.getId(), sessions);
   }
 
   private static SessionExec createSessionExec(
       String sessionId,
       Session session,
-      Map<String, TaskTemplate> idToTaskTemplate,
+      InternalLibrary internalLibrary,
       ExperimentConfig experimentConfig,
       Map<String, Integer> taskTemplateIdToPermuteOrderCounter,
       Map<String, Integer> taskTemplateIdToParameterValuesCounter) {
-    List<TaskExec> tasks = new ArrayList<>();
-    for (int j = 0; j < session.getTasks().size(); j++) {
-      Task task = session.getTasks().get(j);
+    List<Task> tasks;
+    if (session.getTasks() != null) {
+      tasks = session.getTasks();
+    } else if (session.getTemplateId() != null) {
```

Comment:
I have refactored the code a bit here. I believe you would normally solve this with inheritance, that seems the right way. However, the problem is that it required rewriting a lot of code if we did not want to explicitly identify each of the input options, e.g., creating custom serializers/deserializers, etc. I think such effort is not worth it right now, especially since these are objects that are only used while parsing the input files, i.e., once we go through `BenchmarkObjectFactory`, we actually have our internal representation that is the only one that we rely upon while executing experiments. Maybe we can refactor some of this code in the future.